src/Core/Checkout/Payment/PaymentProcessor.php

```diff
@@ -91,43 +106,71 @@ public function pay(
 
             throw $e;
         } finally {
-            if ($token) {
-                // has been nulled, if response is RedirectResponse, therefore we have a finalize step
-                $this->tokenFactory->invalidateToken($token);
+            // has been nulled, if response is RedirectResponse, therefore we have a finalize step
+            if ($encodedToken) {
+                if (($token ?? null) instanceof PaymentToken && $token->jti !== null) {
+                    $this->paymentTokenLifecycle->invalidateToken($token->jti);
+                } else {
+                    $this->tokenFactory->invalidateToken($encodedToken);
+                }
             }
         }
     }
 
-    public function finalize(TokenStruct $token, Request $request, SalesChannelContext $context): TokenStruct
+    /**
+     * @deprecated tag:v6.8.0 - reason:parameter-type-change - first parameter will become `PaymentToken $token` instead of being the last optional parameter
+     * @deprecated tag:v6.8.0 - reason:return-type-change - will return `void` instead of `TokenStruct`
+     *
+     * new signature to copy: public function finalize(PaymentToken $token, Request $request, SalesChannelContext $context): void
+     */
+    public function finalize(TokenStruct $token, Request $request, SalesChannelContext $context /* , ?PaymentToken $paymentToken = null */): TokenStruct
```

Comment:
I got another idea, how the BC could probably be better handled:

```suggestion
    /**
     * @param TokenStruct|PaymentToken $token @deprecated tag:v6.8.0 - reason:parameter-type-change - Will only allow `PaymentToken` as native type
     * @deprecated tag:v6.8.0 - reason:return-type-change - will return `void` instead of `TokenStruct`
     *
     * new signature to copy: public function finalize(PaymentToken $token, Request $request, SalesChannelContext $context): void
     */
    public function finalize($token, Request $request, SalesChannelContext $context): TokenStruct
```

removing the type should not break child classes.
in the method you then can check, which one you got, (if none of them, exception of course :grimacing:) 

with that, callers can already be forward compatible

Reply:
But it breaks decoration: https://3v4l.org/fTOOi
You cannot not define a type and then have one in the extending child class.

Reply:
damn, I thought removing the type would be okay :see_no_evil: